### PR TITLE
Support MIMO refit through model-declared modules

### PR DESCRIPTION
**Pull request by wdykas**
I don't know if this is the best way to do this. I have rewritten this 5 times but MiMO is implemented outside of megatron normal practice  that all my solutions come out slightly confusing.   This is the best I can do for now, input is appreciated..............

- [ ] I, the PR author, have personally reviewed every line of this PR.

# What does this PR do?

Allow the existing prepare/swap API to refit MIMO training models into ordinary Megatron LLaVA inference models on disjoint GPUs. MIMO adds two pieces of information that ordinary parameter iteration does not provide: each component's owning process groups and its inference matching prefix.

An optional model `refit_modules()` method declares `(matching_prefix, live_module, process_group_collection)` tuples. The existing extractor builds metadata using those groups and preserves native storage paths. Rank offsets, pipeline/expert numbering, planning and execution stay in refit. Declarations are consumed during preparation; metadata fingerprints are memoized so warm swaps do not enumerate components again.

MIMO declares its language model, image encoder and projector. The planner validates complete tensor coverage and unique matching names. Projectors retain supplied groups, tied-embedding aliases retain their component prefix, and buffer dtypes match through existing transfer IDs. Composite plan keys use fingerprints of existing planner metadata. Equivalent rebuilt models reuse plans; a weak-key memo of fingerprints releases dead models and clears with the plan cache. Buffer dtype conflicts are checked for every destination on every rank using the existing gather. Public prepare/swap signatures, package exports, checkpoint keys and ordinary model behavior remain unchanged. No registry, component schema, base-model hook or example script is added.

### Review follow-up

- 193 unit tests passed, including equivalent-model cache reuse, cache invalidation, native EP ranks and missing-group diagnostics.
- 67 distributed configurations passed (two eight-GPU cases skipped; 31 NVSHMEM/M2N cases deselected). The eight MIMO cases now rebuild training models while inference retains its object; a new four-rank regression verifies identical dtype-conflict errors on both sources, the destination and an idle rank. All nine MIMO/error cases were rerun after the final fingerprint-cache invalidation adjustment.
- EP2-to-EP1 transfer passed on NCCL. Actual NeMo worker refit across independent worlds passed on Gloo with TP2 dynamic inference. MIMO/DDP scheduling and training, worker refit, dynamic inference and the independent full-batch gradient oracle passed on NCCL.
- Formatting and lint checks passed; the 17 non-blocking mypy diagnostics match the previous baseline.

### Earlier validation

Tested on four GB200 GPUs against upstream `7a9de9c6a`:

- **183 unit tests passed** across metadata, planner, cache, execution, copy services and DP routing.
- **66 distributed configurations passed**: eight MIMO-to-LLaVA cases plus 58 existing GPT/MoE cases. The committed MIMO oracle snapshots source state before transfer and checks source preservation and every destination tensor. Two eight-GPU cases were skipped; 31 NVSHMEM/M2N cases were deselected.
- **17 training/static-inference configurations passed 52 refit rounds**, checking exact state, image-conditioned logits, and cached prefill/decode against uncached full forwards. Coverage includes TP shrink/expansion, tied embeddings, projectors, GQA, FP16/BF16, wrappers, execution batching and cache rebuilds.
- **Nine dynamic-inference configurations passed 27 refit rounds**, including distinct images, different request lengths, reversed insertion order and two cached decode steps at TP1/TP2.
- **Actual MIMO scheduling, heterogeneous DDP and MimoOptimizer training passed** with language DP2 and vision DP1, followed by NeMo-RL's actual worker refit across independent worlds on NCCL/Gloo. Source and inference logits matched exactly. An independent ordinary LLaVA full-batch gradient check passed both optimizer steps: vision/projector gradients were exact; language relative L2 error was about 0.16% under BF16 reduction ordering.
- Additional state-transfer probes passed for component PP/DP/CP ownership, destination PP2, MoE expert TP and EP2-to-EP1, and an opt-in Bridge wrapper subclass declaring its inner LLaVA. Actual worker methods also passed independent worlds of sizes 3+1 and 2+2, including TP2 dynamic inference.
- Black, isort, pylint and Ruff passed. The non-blocking mypy step reports the same 17 diagnostics as the base, with no new diagnostics.

Committed GPU regression command:

```bash
uv run python -m torch.distributed.run --standalone --nproc_per_node=4 -m pytest -q \
  tests/unit_tests/resharding/test_mimo_model_swap.py \
  tests/unit_tests/resharding/test_model_swap.py::test_swap_gpt_parametrized \
  tests/unit_tests/resharding/test_model_swap.py::test_router_expert_bias_refit \
  tests/unit_tests/resharding/test_model_swap.py::test_router_expert_bias_refit_non_collocated \
  -k 'not nvshmem and not nccl_m2n'
```

## Scope

The MIMO mapping targets equivalent core CLIP/projector/language components. Validation uses tiny random models and synthetic images. Different encoder architectures or wrapper names may need their own declarations or weight conversion. FP8, FP4, Kitchen and per-module quantization recipes are rejected. PP/CP/EP ownership probes validate transfer, not complete training/inference for those layouts. Full NeMo-RL/Ray recipe integration, asynchronous engine scheduling and CUDA graphs are outside this change.

Two existing issues were isolated during testing. Static VLM batch-two decoding computes an incorrect image cache offset; it reproduces with plain LLaVA without refit, while the tested dynamic path passes. Router buffers can promote to FP32 after the first transfer caches their earlier dtype; this also reproduces on the base. The README documents initialization to runtime buffer dtypes before first refit, or collective cache clearing and preparation after a dtype change that preserves the tensor object. MoE probes follow that lifecycle. Neither issue adds unrelated fixes to this PR.

## Contribution process

- [x] Added relevant unit tests, including distributed real-model refit tests.
- [ ] Added training-recipe functional tests.
- [x] Added documentation.
- [x] Ran repository autoformat checks; non-blocking mypy baseline noted above.

### CODE REVIEW

**Comment by wdykas** [comment]
/ok to test [715d0b5](https://github.com/NVIDIA/Megatron-LM/commit/715d0b5fadc1b4a4c7d3255e55704d7adc6ecb43)

**Comment by wdykas** [comment]
/ok to test [e784d2f](https://github.com/NVIDIA/Megatron-LM/pull/7367/commits/e784d2fc1cb5b5c0eeee3c0e18dc5fc9a931ab00)

File: megatron/core/models/mimo/model/base.py [hunk 1]
```diff
@@ -100,6 +103,63 @@ def __init__(self, mimo_config: MimoModelConfig, cp_group=None, tp_group=None) -
...
+                    )
+                )
+
+        for label, module, pg in components:
+            if pg is None or any(getattr(pg, axis, None) is None for axis in ("tp", "pp", "dp")):
+                raise ValueError(
+                    f"Refit component {label!r} requires explicit tp, pp and dp groups"
+                )
+            config = getattr(module, "config", None)
+            if (
```
Comment (wdykas) [inline megatron/core/models/mimo/model/base.py:145]: gotta add support here if we care

**Comment by cspades** [comment]
Starting to pipeclean using an AI-generated MiMo RL pipeline (using grpo.py so should be a sufficient test for something that doesn't exist yet).

Will approve when I finish!

**Comment by cspades** [comment]
Okay getting a consistently low generation KL error with this mock MiMo setup using TP2/EP2 -> TP1/EP1 for generation:

- https://github.com/cspades/RL/tree/cye/mimo-mllm-refit-plumb
- https://github.com/cspades/Megatron-Bridge/tree/cye/nemotron-omni-super-v3-mimo-dev
- https://github.com/cspades/Megatron-LM/tree/cye/mlm-tot-vlm-rebase-mimo

Minimal changes in MBridge and MCore that we'll probably have to pipeclean for MiMo support in RL.

```
📊 Training Results:
  • Loss: 0.0000
  • Generation KL Error: 0.0019
  • Avg Reward: 0.0000
  • Mean Generation Length: 64.0000
```
Even with no reward (since the model is a randomly-initialized, truncated Omni-LLaVA), whatever logits the generation model is producing is within the magnitude of the policy that I see for the CLEVR image dataset, and I also added some debug code where I write `NaN` into the target to make sure the refit updates it.

We'll need to revisit this again when MiMo is properly supported in RL to be safe.

**Review by cspades — APPROVED** [review]
Able to make this PR work as-is with the MiMo RL support I hacked up.

**Review by kvareddy — APPROVED** [review]

**Review by santhnm2 — APPROVED** [review]

File: megatron/core/models/mimo/model/base.py (hunk 1 above)
Comment (cspades) [reply megatron/core/models/mimo/model/base.py:145]: Yeah, let's defer to a future PR. This one will support BF16 inference.

**Review by Phlip79 — APPROVED** [review]
